### PR TITLE
Add sandbox

### PR DESCRIPTION
**Pull request by markusjura**
This PR is adding the sandbox to the conductr-cli.

### CODE REVIEW

**Comment by markusjura** [comment]
This PR is still work in progress. 

**Done**
- CLI parser for `sandbox run` and `sandbox stop` (beside some cosmetics)
- Implementation of `sandbox run`

*ToDo*
- Implementation of `sandbox stop`
- Implementation of `sandbox debug` (mostly reuse of `sandbox run` functions)
- CLI parser for `sandbox debug (reuse of `sandbox run` parser)
- Tests

**Comment by markusjura** [comment]
@huntc: Can you do a first review?

**Comment by huntc** [comment]
Sure thing. Excited.

File: conductr_cli/sandbox.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import argparse
+import os
+from conductr_cli import sandbox_start, sandbox_stop
+
+DEFAULT_IP = os.getenv('CONDUCTR_IP', '127.0.0.1')
+
+def add_default_arguments(sub_parser):
+    sub_parser.add_argument('-i', '--ip',
+                            default=DEFAULT_IP,
+                            help='ConductR IP address. Defaults to $CONDUCTR_IP or 127.0.0.1.')
```
Comment (huntc) [inline conductr_cli/sandbox.py]: Not sure that I understand why you'd specify an IP for the sandbox...

File: conductr_cli/sandbox.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                            metavar='')
+    run_parser.add_argument('-m', '--image',
+                            default='typesafe-docker-registry-for-subscribers-only.bintray.io/conductr/conductr-dev',
+                            help='Docker image to use.\n'
+                                 'Defaults to `typesafe-docker-registry-for-subscribers-only.bintray.io/conductr/conductr`.',
+                            metavar='')
+    run_parser.add_argument('-v', '--image-version',
+                            default='1.0.11',
+                            help='Version of the Docker image to use.\n'
+                                 'Defaults to 1.0.11.\n'
```
Comment (huntc) [inline conductr_cli/sandbox.py]: 1.0.12 is released, and the Docker image for it should be available in less than 24 hours. Meanwhile, I suggest changing this now, and tagging a 1.0.12 in your local registry.

File: conductr_cli/sandbox.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    argcomplete.autocomplete(parser)
+    args = parser.parse_args()
+    if vars(args).get('func') is None:
+        parser.print_help()
+    else:
+        args.func(args)
+
+
+if __name__ == '__main__':
+    run()
```
Comment (huntc) [inline conductr_cli/sandbox.py:128]: Reallly nice

File: conductr_cli/sandbox_start.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+#############################
+CONDUCTR_DEV_IMAGE = 'typesafe-docker-registry-for-subscribers-only.bintray.io/conductr/conductr-dev'
+LATEST_CONDUCTR_VERSION = '1.0.11'
+CONDUCTR_NAME_PREFIX = 'cond-'
+CONDUCTR_PORTS = { 5601, # conductr-kibana bundle
+                   9004, # ConductR internal akka remoting
+                   9005, # ConductR controlServer
+                   9006, # ConductR bundleStreamServer
+                   9200, # conductr-elasticsearch bundle
+                   9999} # visualizer bundle
```
Comment (huntc) [inline conductr_cli/sandbox_start.py]: sp before }

File: conductr_cli/sandbox_start.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
+from conductr_cli import terminal
+
+
+#############################
+###       Constants       ###
+#############################
```
Comment (huntc) [inline conductr_cli/sandbox_start.py]: Have we been using this commenting convention elsewhere? If not then either we apply globally, or not at all.

File: conductr_cli/sandbox.py (hunk 2 above)
Comment (markusjura) [inline conductr_cli/sandbox.py]: Sure, just wanted to test that yesterday evening..

File: conductr_cli/sandbox_start.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+def run(args):
+    """`sandbox run` command"""
+    # TODO: pull_image(args)
+    ports = collect_ports(args)
+    scale_cluster(args, ports)
+
+def debug(args):
+    """`sandbox debug` command"""
+    print("TODO: Implement")
```
Comment (huntc) [inline conductr_cli/sandbox_start.py]: I'm starting to wonder about the utility of debugging outside of an build environment such as sbt. WDYT?

File: conductr_cli/sandbox_start.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        port_args.append(map_port(instance, port))
+    return port_args
+
+def resolve_docker_run_positional_args(cond0_ip):
+    seed_node_arg = ['--seed', '{}:{}'.format(cond0_ip, 9004)] if cond0_ip else []
+    return ['--discover-host-ip'] + seed_node_arg
+
+# Stop nodes
+def stop_nodes():
+    '''TODO: Implement'''
```
Comment (huntc) [inline conductr_cli/sandbox_start.py]: Just wondering if you should seek any inspiration from the original run_conductrs python script we had - perhaps not, but just thought I'd remind you of it.

File: conductr_cli/sandbox.py (hunk 1 above)
Comment (markusjura) [inline conductr_cli/sandbox.py]: We need that to expose the {ip}:{port} when starting the sandbox.

Generally I was thinking about implementing the same `default ip` logic into the `conductr-cli` as we have it in `sbt-conductr`:
- Try to get ip from `docker-machine`, else
  - Try to get ip from `boot2docker`, else
    - Try to get ip from `hostname`, else
      - 127.0.0.1

WDYT?

File: conductr_cli/terminal.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+def docker_inspect(id, format = None):
+    format_arg = ['--format', format] if format else []
+    command = ['docker', 'inspect'] + format_arg + [id]
+    return subprocess.check_output(command, universal_newlines=True).strip()
+
+def docker_run(optional_args, image, positional_args):
+    command = ['docker', 'run'] + optional_args + [image] + positional_args
+    print('docker command: {}'.format(' '.join(command)))
+    return subprocess.call(['docker', 'run'] + optional_args + [image] + positional_args)
```
Comment (huntc) [inline conductr_cli/terminal.py]: Is console a better name than terminal?

File: conductr_cli/sandbox_start.py (hunk 4 above)
Comment (markusjura) [inline conductr_cli/sandbox_start.py]: Done

**Comment by markusjura** [comment]
**Update**
- Review comments have been resolved / implemented
- Logic to automatically resolve host IP address with `docker-machine`, `boot2docker`, `hostname` has been implemented
- `sandbox` parser now checks if the `image_version` has been set. Otherwise the program exits with a message that the version need to be obtained from typesafe.com (same message as in `sbt-conductr-sandbox`)
- `sandbox` parser has been optimized
- `sandbox debug` command has been implemented (disabled for now, as discussed with @huntc)
- `sandbox stop` command has been implemented
- `conductr` and `sandbox` parser have been aligned (same help output)
- README.rst has been updated
- Tests are still missing

In summary the new `sandbox` command fully works with the conductr image `conductr/conductr` and the version `1.0.12`, but no tests have been written yet.

**Comment by markusjura** [comment]
Implementation is done, smoke tests on linux, OS X and Windows have been successful.  Only unit tests are still missing.

**Today's progress**
- Added latest python3 version to test config (currently Python 3.5.0)
- Fixed `conduct` command
- Fixed sandbox port mapping logic
- Improved `sandbox` error handling for linux
- Fixed the `sandbox --feature` option
- Smoke tests on OS X and linux (sandbox run, stop, run with features, conduct info, load, run, stop)

**Comment by huntc** [comment]
Tests on CI now pass also. Awesome. I can't wait to try this.

**Comment by edwardcallahan** [comment]
Looking good. Tested visualization, multi container, service lookup, general bundle usage on Linux. Nice work, Markus.

**Comment by huntc** [comment]
If we attempt to run a `conduct` command and there is no ConductR running, is it reasonable to also suggest starting the sandbox (bearing in mind we only do so when --ip points to a VM)? Presently:

```
Christophers-MacBook-Pro:typesafe-conductr-cli huntc$ python3 -m conductr_cli.conduct services
ERROR: Unable to contact ConductR.
ERROR: Reason: HTTPConnectionPool(host='192.168.59.103', port=9005): Max retries exceeded with url: /bundles (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x10a868e80>: Failed to establish a new connection: [Errno 61] Connection refused',))
ERROR: Make sure it can be accessed at http://192.168.59.103:9005/bundles
```

**Comment by huntc** [comment]
It'd be nice to see an example of setting the roles in the help text.

**Comment by huntc** [comment]
Should we align sbt-sandbox's --with-feature with the sandbox here i.e. --feature for each feature...?

**Comment by markusjura** [comment]
Created an issue to align the arguments: https://github.com/typesafehub/sbt-conductr-sandbox/issues/63

**Comment by markusjura** [comment]
> It'd be nice to see an example of setting the roles in the help text.

Yes, will add this.

**Comment by markusjura** [comment]
Introduced a new argument `--local-connection` which is not displayed in the help. This argument is true if the `--ip` argument is not getting overriden in the `conduct` command. For the `sandbox` commands `--local-connection` is always True.

If `--local-connection` is True then the error message is different if ConductR can't be reached:

```
mj@MacBook-Pro-2:~/workspace/conductr-cli$ python3 -m conductr_cli.conduct info
ERROR: Unable to contact ConductR.
ERROR: Reason: HTTPConnectionPool(host='192.168.99.100', port=9005): Max retries exceeded with url: /bundles (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x1101e75c0>: Failed to establish a new connection: [Errno 61] Connection refused',))
ERROR: Start the ConductR sandbox with: sandbox run IMAGE_VERSION
```

The last line changed accordingly pointing the user towards how to start the sandbox.

**Comment by huntc** [comment]
When would a user need to use the local connection arg?

**Comment by markusjura** [comment]
Never, it is basically a hidden argument which the user does not see in the help and which we will not document. We need this flag to check during error handling if the connection is a local or remote connection, i.e. if the user has specified the `--ip` arg. During error handling we can't check this based on the `--ip` argument anymore because the value is set to an ip address. Therefore we need an additional argument `--local-connection`.

**Comment by huntc** [comment]
Feels then like we are overloading the use of args when what we are trying to do is set an internal flag... WDYT?

**Comment by markusjura** [comment]
I think it is fine. In the `argparse` library there is a notation to `suppress` things which we use in this case:

```
sub_parser.add_argument('--local-connection',
  default=True,
  help=argparse.SUPPRESS)
```

So it seems to be OK. Also if we would not put this flag into the args object then it would make things much more complicated. We couldn't use `parser.set_defaults()` anymore and at the end call `args.func(args)`.

**Comment by markusjura** [comment]
Will push the tests on Sunday and then merge this PR on Sunday as well.

**Comment by huntc** [comment]
Thanks Markus! Happy to review the tests also. Btw the tests are presently failing on this PR.

**Comment by markusjura** [comment]
Tests has been added.

@huntc: Can you please do a final review?

**Comment by huntc** [comment]
Commits need to be squashed.

File: conductr_cli/sandbox.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                                 'In this example the roles role1 and role2 are assigned to the first node and\n'
+                                 'role3 is assigned to the second node.\n'
+                                 'Afterwards the specified roles are subsequently applied to the remaining nodes.\n'
+                                 'Defaults to [].',
+                            metavar='')
+    run_parser.add_argument('-d', '--debug-port',
+                            type=int,
+                            default=5005,
+                            help='Debug port to be made public by each of the ConductR containers. Defaults to 5005.',
+                            metavar='')
```
Comment (huntc) [inline conductr_cli/sandbox.py]: Is this arg still required?

**Comment by huntc** [comment]
Other than https://github.com/typesafehub/conductr-cli/pull/73#issuecomment-155232196 and https://github.com/typesafehub/conductr-cli/pull/73#discussion_r44351682, LGTM. Please address those two things and then merge/release.

Thanks for the awesomeness!

File: conductr_cli/sandbox.py (hunk 9 above)
Comment (markusjura) [inline conductr_cli/sandbox.py]: No, it's not. Good catch.

**Comment by markusjura** [comment]
Mistakenly renamed the branch and therefor PR got closed.

**Comment by markusjura** [comment]
Commits have been squashed.
